wrangler-service/src/main/java/io/cdap/wrangler/service/directive/WorkspaceHandler.java

```diff
@@ -222,6 +218,52 @@ public void updateWorkspace(HttpServiceRequest request, HttpServiceResponder res
     });
   }
 
+  /**
+   * Resample the workspace using a new sample request
+   */
+  @POST
+  @TransactionPolicy(value = TransactionControl.EXPLICIT)
+  @Path("v2/contexts/{context}/workspaces/{id}/resample")
+  public void resampleWorkspace(HttpServiceRequest request, HttpServiceResponder responder,
+                                @PathParam("context") String namespace,
+                                @PathParam("id") String workspaceId) {
+    respond(responder, namespace, ns -> {
+      if (ns.getName().equalsIgnoreCase(NamespaceId.SYSTEM.getNamespace())) {
+        throw new BadRequestException("Resampling workspace in system namespace is currently not supported");
+      }
+
+      WorkspaceResampleRequest resampleRequest =
+        GSON.fromJson(StandardCharsets.UTF_8.decode(request.getContent()).toString(), WorkspaceResampleRequest.class);
+
+      WorkspaceId wsId = new WorkspaceId(ns, workspaceId);
+      Workspace currentWorkspace = wsStore.getWorkspace(wsId);
+
+      String connectionName = currentWorkspace.getSampleSpec() == null ? null :
+        currentWorkspace.getSampleSpec().getConnectionName();
+      if (connectionName == null) {
+        throw new BadRequestException("Connection name has to exist to resample a workspace");
+      }
+
+      SampleRequest sampleRequest = resampleRequest.getSampleRequest();
+      if (sampleRequest == null) {
+        throw new BadRequestException("Sample request has to be provided to resample a workspace");
+      }
+
+      SampleResponse sampleResponse = discoverer.retrieveSample(namespace, connectionName,
+                                                                sampleRequest);
+      List<Row> rows = getSample(sampleResponse);
+
+      SampleSpec newSpec = currentWorkspace.getSampleSpec();
+      newSpec.setSampleRequest(sampleRequest);
```

Comment:
don't use setter, if you want to copy, just use the constructor

Reply:
Good point -- fixed!